### PR TITLE
docs/examples: Fix and add docs for Wasm C++ sandbox

### LINKED ISSUE #13605: Add docs for Wasm Sandbox
### description

We have a sandbox to demonstrate creating a Wasm filter, but it doesnt currently have any documentation

### PR DESCRIPTION
**Pull request by phlax**
Signed-off-by: Ryan Northey <ryan@synca.io>

Commit Message: docs/examples: Fix and add docs for Wasm sandbox
Additional Description:
- provide a pre-compiled wasm binary for example
- provide recipe for updating wasm
- fix verify.sh and composition permissions
- add docs

Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue] Fix #13605 
[Optional Deprecated:]

### CODE REVIEW

**Comment by phlax** [comment]
the sandbox here pretty much works. Ill update the docs next...

it includes a routine which:
- brings up a composition with a pre-compiled wasm
- checks content/headers
- brings down composition
- recompiles a new wasm
- brings back up and rechecks

i think this is very helpful for end users of the sandbox but it adds 5mins+ to the CI

File: examples/wasm/docker-compose-wasm.yaml [hunk 1]
```diff
@@ -0,0 +1,23 @@
+version: "3.7"
+
+services:
+
+  wasm_compile_update:
+    image: envoyproxy/envoy-build-ubuntu:b480535e8423b5fd7c102fd30c92f4785519e33a
```
Comment (phlax) [inline examples/wasm/docker-compose-wasm.yaml:6]: its not ideal hardcoding the (commit head sha) like this, im wondering if we can add a `latest` tag  on this dockerhub repo

File: examples/wasm/BUILD [hunk 2]
```diff
@@ -8,10 +8,31 @@ licenses(["notice"])  # Apache 2
 
 envoy_package()
 
+filegroup(
+    name = "configs",
+    srcs = glob(
+        [
+            "**/*.yaml",
+            "**/*.wasm",
```
Comment (phlax) [inline examples/wasm/BUILD:16]: this only works on amd64, im wondering how i can exclude it from other arch builds

Comment (phlax) [reply examples/wasm/BUILD]: this should be fixed

File: examples/wasm/docker-compose.yaml [hunk 3]
```diff
@@ -28,10 +20,8 @@ services:
       envoymesh:
         aliases:
           - web_service
-    expose:
-      - "80"
     ports:
-      - "18080:80"
+      - "9000:9000"
```
Comment (phlax) [inline examples/wasm/docker-compose.yaml:24]: this is probs not necessary

File: examples/wasm/BUILD [hunk 4]
```diff
@@ -8,10 +8,37 @@ licenses(["notice"])  # Apache 2
...
 envoy_package()
 
+filegroup(
+    name = "configs",
+    srcs = glob(
+            [
+                "**/*.wasm",
+            ],
+    ) + select({
+        "//bazel:linux_x86_64": glob(
```
Comment (phlax) [inline examples/wasm/BUILD:18]: this passes `config_validation` on the `x64 release` run, but fails on `compile_time_options` - not sure why

File: .azure-pipelines/pipelines.yml [hunk 5]
```diff
@@ -151,7 +151,7 @@ stages:
               AZP_BRANCH: $(Build.SourceBranch)
 
   - stage: linux_x64
-    dependsOn: ["precheck"]
```
Comment (phlax) [inline .azure-pipelines/pipelines.yml:154]: remove me

File: examples/wasm/BUILD (hunk 4 above)
Comment (phlax) [reply examples/wasm/BUILD:18]: https://dev.azure.com/cncf/envoy/_build/results?buildId=54678&view=logs&j=e969334a-0e55-5c18-ac96-8b546753391e&t=b5329a9e-93c7-5412-9f4c-6f8d659e1b57&l=4144

**Comment by phlax** [comment]
i have some small changes remaining, but docs are rendered here:

https://storage.googleapis.com/envoy-pr/13622/docs/start/sandboxes/wasm.html

File: examples/wasm/BUILD (hunk 4 above)
Comment (phlax) [reply examples/wasm/BUILD:18]: thinking this should actually work on win/mac also - just only on x86_64 arch

File: configs/configgen.sh [hunk 6]
```diff
@@ -10,6 +10,7 @@ shift
 mkdir -p "$OUT_DIR/certs"
 mkdir -p "$OUT_DIR/lib"
 mkdir -p "$OUT_DIR/protos"
+mkdir -p "$OUT_DIR/wasmlib"
```
Comment (phlax) [inline configs/configgen.sh:13]: revert to just using lib

File: examples/wasm/envoy_filter_http_wasm_example.cc [hunk 7]
```diff
@@ -65,8 +65,9 @@ FilterHeadersStatus ExampleContext::onResponseHeaders(uint32_t, bool) {
   for (auto& p : pairs) {
     LOG_INFO(std::string(p.first) + std::string(" -> ") + std::string(p.second));
   }
-  addResponseHeader("newheader", "newheadervalue");
+  addResponseHeader("X-Wasm-custom", "FOO");
   replaceResponseHeader("location", "envoy-wasm");
```
[the remark this answers is not in the document]
Comment (phlax) [reply examples/wasm/envoy_filter_http_wasm_example.cc:69]: yep - it was the example provided - ill update

**Comment by phlax** [comment]
> compile_time_options doesn't include any Wasm runtimes.

this is for `config_validation` only

File: examples/wasm/BUILD (hunk 4 above)
Comment (phlax) [reply examples/wasm/BUILD:18]: ah, k - i see

not sure how to exclude this test

File: docs/root/start/sandboxes/wasm.rst [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+
+Stop the proxy server and compile the Wasm binary with the updated file:
+
+.. code-block:: console
+
+   $ docker-compose stop proxy
+   $ docker-compose -f docker-compose-wasm.yaml up --remove-orphans wasm_compile_update
+
+
+Step 4: Check the updated web response
```
Comment (phlax) [inline docs/root/start/sandboxes/wasm.rst:79]: step 5

File: examples/wasm/BUILD (hunk 4 above)
Comment (phlax) [reply examples/wasm/BUILD]: i think this should be resolved now

File: docs/root/start/sandboxes/wasm.rst [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+edit...
+
+Restart the proxy server
+
+.. code-block:: console
+
+   $ docker-compose up --build -d proxy
+
+
+The Wasm filter should now inject "Hello, Wasm world" into the end of the response body.
```
Comment (phlax) [inline docs/root/start/sandboxes/wasm.rst:93]: s/into/at/

File: docs/root/start/sandboxes/wasm.rst [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+   $ docker-compose stop proxy
+   $ docker-compose -f docker-compose-wasm.yaml up --remove-orphans wasm_compile_update
+
+
+Step 4: Check the updated web response
+**************************************
+
+Edit the Docker recipe to use the updated binary
+
+edit...
```
Comment (phlax) [inline docs/root/start/sandboxes/wasm.rst:84]: ...edit

File: docs/root/start/sandboxes/wasm.rst [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+.. code-block:: console
+
+   $ docker-compose up --build -d proxy
+
+
+The Wasm filter should now inject "Hello, Wasm world" into the end of the response body.
+
+.. code-block:: console
+
+   $ curl -s http://localhost:8000 | grep "Hello, world"
```
Comment (phlax) [inline docs/root/start/sandboxes/wasm.rst:97]: s/ world/ Wasm world/

**Comment by phlax** [comment]
@PiotrSikora this should be good for review

**Comment by phlax** [comment]
couple of thoughts ive had about this PR

- i removed the admin interface from config, not sure if there is something we should show/test with that
- im wondering if we should rename this to "C++ Wasm Filter" and then add a "Go Wasm Filter" also

File: examples/wasm/docker-compose-wasm.yaml (hunk 1 above)
Comment (phlax) [reply examples/wasm/docker-compose-wasm.yaml:6]: afaict there is no intention to resolve this - so afaict the only way to document this is hardcoding it - i think anything else is outside the scope of this PR

**Comment by phlax** [comment]
I had a play around with the `go` and `rust` sdks to see if i could easily build equivalents to this sandbox.

i didnt quite either to compile, but i got the impression it would not be too hard to get both working.

ive therefore renamed this sandbox to `wasm-cc` and adjusted the docs so that we can more easily add those other sandboxes after

**Comment by phlax** [comment]
this has started failing on `clang_tidy` not sure why - but i guess related to renaming the example. 

i tried running clang_tidy outside of CI on this branch, and it passed so im a bit confused as to the reason

/retest

**Review by mattklein123 — APPROVED** [review]
Very cool. Love seeing these working examples.

**Pull request merged**